src/java_bytecode/generic_parameter_specialization_map_keys.cpp

```diff
@@ -0,0 +1,152 @@
+  pairs.reserve(parameters.size());
+  std::transform(
+    parameters.begin(),
+    parameters.end(),
+    types.begin(),
+    std::back_inserter(pairs),
+    [&](java_generic_parametert param, reference_typet type)
+    {
+      return std::make_pair(param, type);
+    });
+
+  for(const auto &pair : pairs)
+  {
+    // Only add the pair if the type is not the parameter itself, e.g.,
+    // pair.first = pair.second = java::A::T. This can happen for example
+    // when initiating a pointer to an implicitly java generic class type
+    // in gen_nondet_init and would result in a loop when the map is used
+    // to look up the type of the parameter.
+    if(
+      !(is_java_generic_parameter(pair.second) &&
+        to_java_generic_parameter(pair.second).get_name() ==
+          pair.first.get_name()))
+    {
+      const irep_idt &key = pair.first.get_name();
+      if(generic_parameter_specialization_map.count(key) == 0)
+        generic_parameter_specialization_map.emplace(
+          key, std::stack<reference_typet>());
+      (*generic_parameter_specialization_map.find(key))
+        .second.push(pair.second);
+
+      // We added something, so pop it when this is destroyed:
+      erase_keys.push_back(key);
+    }
+  }
+}
+
+/// Add pairs of parameters and their types for a generic pointer type to the
+/// controlled map. Own the keys and pop from their stack on destruction;
+/// otherwise do nothing.
+/// \param pointer_type pointer type to get the specialized generic types from
+/// \param pointer_subtype_struct struct type to which the generic pointer
+/// points, must be generic if the pointer is generic
+const void generic_parameter_specialization_map_keyst::insert_pairs_for_pointer(
+  const pointer_typet &pointer_type,
+  const typet &pointer_subtype_struct)
+{
+  if(is_java_generic_type(pointer_type))
+  {
+    // The supplied type must be the full type of the pointer's subtype
+    PRECONDITION(
+      pointer_type.subtype().get(ID_identifier) ==
+      pointer_subtype_struct.get(ID_name));
+
+    const java_generic_typet &generic_pointer =
+      to_java_generic_type(pointer_type);
+
+    // If the pointer points to an incomplete class, don't treat the generics
+    if(!pointer_subtype_struct.get_bool(ID_incomplete_class))
+    {
+      PRECONDITION(
+        is_java_generic_class_type(pointer_subtype_struct) ||
+        is_java_implicitly_generic_class_type(pointer_subtype_struct));
+      const std::vector<java_generic_parametert> &generic_parameters =
+        get_all_generic_parameters(pointer_subtype_struct);
+
+      INVARIANT(
+        generic_pointer.generic_type_arguments().size() ==
+          generic_parameters.size(),
+        "All generic parameters of the pointer type need to be specified");
+
+      insert_pairs(
+        generic_parameters, generic_pointer.generic_type_arguments());
+    }
+  }
+}
+
+/// Add pairs of parameters and their types for a generic symbol type to the
```

Comment:
Perhaps improve this comment by explaining that the pairs are for each generic parameter and why you would use this overload of the method vs. the above one (e.g. something about being used for the base class?)